java/org/apache/catalina/util/TimeBucketCounter.java

```diff
@@ -0,0 +1,217 @@
+     * bit shift arithmetic
+     *
+     * @return the actual bucket duration in milliseconds
+     */
+    public int getActualDuration() {
+        return (int) Math.pow(2, getNumBits());
+    }
+
+    /**
+     * returns the ratio between the configured duration param and the
+     * actual duration which will be set to the next power of 2.  we then
+     * multiply the configured requests param by the same ratio in order
+     * to compensate for the added time, if any
+     *
+     * @return the ratio, e.g. 1.092 if the actual duration is 65_536 for
+     *         the configured duration of 60_000
+     */
+    public double getRatio() {
+        return ratio;
+    }
+
+    /**
+     * returns the ratio to the next power of 2 so that we can adjust the value
+     *
+     * @param value
+     * @return
+     */
+    static double ratioToPowerOf2(int value) {
+        double nextPO2 = nextPowerOf2(value);
+        return Math.round((1000 * nextPO2 / value)) / 1000d;
+    }
+
+    /**
+     * returns the next power of 2 given a value, e.g. 256 for 250,
+     * or 1024, for 1000
+     *
+     * @param value
+     * @return
+     */
+    static int nextPowerOf2(int value) {
+        int valueOfHighestBit = Integer.highestOneBit(value);
+        if (valueOfHighestBit == value)
+            return value;
+
+        return valueOfHighestBit << 1;
+    }
+
+    /**
+     * when we want to test a full bucket duration we need to sleep until the
+     * next bucket starts
+     *
+     * @return the number of milliseconds until the next bucket
+     */
+    public long getMillisUntilNextBucket() {
+        long millis = System.currentTimeMillis();
+        long nextTimeBucketMillis = ((millis + (long) Math.pow(2, numBits)) >> numBits) << numBits;
+        long delta = nextTimeBucketMillis - millis;
+        return delta;
+    }
+
+    /**
+     * sets isRunning to false to terminate the maintenance thread
+     */
+    public void destroy() {
+        this.isRunning = false;
+    }
+
+    /**
+     * this class runs a background thread to clean up old keys from the map
+     */
+    class MaintenanceThread extends Thread {
+
+        final long sleeptime;
+
+        public MaintenanceThread(long sleeptime) {
+            super.setDaemon(true);
+            this.sleeptime = sleeptime;
+        }
+
+        @Override
+        public void start() {
+            isRunning = true;
+            super.start();
+        }
+
+        @Override
+        public void run() {
+
+            while (isRunning) {
+                String currentBucketPrefix = String.valueOf(getCurrentBucketPrefix());
+                ConcurrentHashMap.KeySetView<String, AtomicInteger> keys = map.keySet();
+
+                for (String k : keys) {
+                    if (!k.startsWith(currentBucketPrefix)) {
+                        // the key is obsolete, remove it
+                        keys.remove((k));
+                    }
+                }
```

Comment:
Here can be replaced with `keys.removeIf(k -> !k.startsWith(currentBucketPrefix));`, which is more streamlined.

Reply:
It depends upon how far back you want to back-port. Tomcat 8.5.x runs on Java 7. Though we build using Java 11, I think this should be buildable on Java 7 as well. Feel free to use lambdas in 9.0+ but please use the long-form in 8.5.x.

Reply:
Looks like Collection.removeIf() has been added in Java 8 [1].  The code also makes use of Map.computeIfAbsent() which was also add in Java 8.

Since Tomcat 8.5 EOL is exactly a year from now [2], we need to decide whether we should port this back to 8.5.  The main question that I have to Tomcat 8.5 users is: what's holding you back from upgrading to Tomcat 9.0?

[1] https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#removeIf-java.util.function.Predicate-
[2] https://tomcat.apache.org/tomcat-85-eol.html

Reply:
Inertia.

Feel free to just _not_ back-port to 8.5.x.

Reply:
Chris - Let's see if the Filter gets any traction.  If we notice from mailing list conversations etc. that it is popular enough then we can port it back to 8.5.x with the small required changes.